### PR TITLE
[TE4] Enable Lighting Devices types cluster and attributes

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
#### Problem
Color control cluster is not enable and is incomplete
Occupancy sensor cluster is not enable

#### Change overview
Enable occupancy Sensor cluster to Controllers, All-cluster-app and Lighting-app example
Enable Color control cluster to Controllers, All-cluster-app and Lighting-app example
Process EnhancedHue commands received
Update xml to match spec (missing arguments in command)
Include Color-control.cpp to Nordic and Telink lighting example

NOTE : there is 2 commits that are only regenerated files. You can focus the review on the two others.

**TODO** 
- ColorLoop not implemented
- Apply Level and/or color to a Interface (Led/Lightbulb)

#### Testing
Rebuild chip-device-ctrl
Complete commissioning flow with EFR32
Test cluster commands 
    zcl LevelControl
    zcl ColorControl
Validate value sent with logs and verify attributes values changes with
   zclread LevelControl CurrentLevel
   zclread ColorControl CurrentHue
   zclread ColorControl CurrentX
   zclread ColorControl CurrentY

### CODE REVIEW

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1928]: Again, it's not clear to me why this can't just do subtraction on `uint16_t` values and cast to `uint16_t`.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (jmartinez-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:319]: @bzbarsky-apple Thank you for all your comments. That is really helpful, and it makes more sense. I will fix all of those. Do you know were I could find the details of those low 8 bits to interpolate the hue, I didn't find that. I just assumed the enhancedhue values were deal separately and gave a weird range of value to represent colors when translated to x y colors

Comment (bzbarsky-apple) [reply src/app/clusters/color-control-server/color-control-server.cpp:319]: That's a good question.  Maybe the ZLL specs?  @CamWms do you know where this information is to be found?

**Comment by jmartinez-silabs** [comment]
@bzbarsky-apple I fixed what you pointed out in your prior review and reimplemented the Enhanced color command callbacks to match spec as discussed with Rob. It also remove duplicated code as Standard Hue and Enhanced Hue commands share the same functions with some condtions. 

@andy31415 @Damian-Nordic @msandstedt @chrisdecenzo @hawk248 I would need your review as this needs to be merge for Friday for TE4 branch. Thank you

I will be Out of Office tomorrow as it is Canada day. Hopefully we can get this merged by friday

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:328]: Perhaps document that when `isEnhanced` is true `rate` will be in the `uint8_t` range?

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:522]: Diff might have been cleaner (and easier to review) if the order had been kept the same (as in, if this were up where `emberAfColorControlClusterMoveToHueAndSaturationCallback` used to be).  Not a big deal, I guess.

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:572]:
```suggestion
        colorHueTransitionState.initialEnhancedHue = currentHue;
        colorHueTransitionState.currentEnhancedHue = currentHue;
```

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:578]:
```suggestion
        colorHueTransitionState.initialHue = static_cast<uint8_t>(currentHue);
        colorHueTransitionState.currentHue = static_cast<uint8_t>(currentHue);
```

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:703]: This function did not change at all, so might have been nice to leave it in the old place in the file, but ok.

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1640]: This rearrangement of the arithmetic is not right.

Say we are moving hue from initial 100 to final 200.  The `subtractHue` result is 100.  Let's say we're doing it over 50 steps and we have taken 1 of those steps.  So `stepsTotal` is 50 and `stepsRemaining` is 49.  The old calculation came out with `newHue32 = (100 * 49) / 50 = 4900 / 50 = 98`.  Then we would subtract that from the final value and end up with a current hue value of 102, as expected.  And if we had taken 45 steps and had 5 stepsRemaining, we would get a current hue of 190.

The new calculation does `newHue32 = 100 * (49/50) = 100 * 0 = 0` (integer division, remember!), and comes out with a current hue of 200.  In fact, anything other than stepsRemaining == stepsTotal gives 200.

I wish we had test coverage for this, given all the changes we are making...  :(  I would hope it would have caught this problem.  Can you please make sure (at least as a followup) that tests are created?

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1655]: Again, this rearrangement of what's divided by what completely breaks the code.

File: src/app/zap-templates/zcl/data-model/silabs/zll.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/silabs/zll.xml:189]: Those args are in ZCL8 as far as I can tell.  That applies throughout this file.  So not sure we should claim `introducedIn="matter-spec"`.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
This is much better, thank you!

I'm still marking "Request changes" because of the integer division issue.  That absolutely needs to be fixed before this can be merged, and if I don't do this it's too easy for other people to mark approvals and for this to merge with comments still outstanding.

That said, I am out of office on Friday.  So I'm fine with the "request changes" being dismissed if this issue is addressed, and my review being treated as an approval at that point...

File: src/app/zap-templates/zcl/data-model/silabs/zll.xml
Comment (jmartinez-silabs) [reply src/app/zap-templates/zcl/data-model/silabs/zll.xml]: I changed it to zcl8. I looked but wasn't sure of the exact zcl8 revision file when it was added.

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (jmartinez-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: I see. Thank you for catching this Boris. I splitted them back how they were Multiplication first. then divide that result.

I will take actions to start a test coverage for this cluster starting next week

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (jmartinez-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: Moved it back up where it was. Total diff should be cleaner now

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@jmartinez-silabs Sorry conflicts here :(

**Comment by bzbarsky-apple** [comment]
@mspang @andy31415 @LuDuda @msandstedt @saurabhst @Damian-Nordic @jepenven-silabs Please take a look?

**Review by msandstedt — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Comment by woody-apple** [comment]
@Damian-Nordic @LuDuda ?

**Comment by woody-apple** [comment]
@mspang ?

**Review by woody-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

File: src/app/clusters/color-control-server/color-control-server.cpp
Comment (andy31415) [inline src/app/clusters/color-control-server/color-control-server.cpp:103]: nit: in C++ code we should probably use typesafe constexpr for constants instead of defines

**Comment by andy31415** [comment]
TE4 cherrypick failed with conflicts on

```
Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)
	deleted by us:   src/app/common/gen/callback.h
```

i believe this will compound as we add more files with zap-related changes in them (I see #7183 and #8138 to have potential issues). Need a path forward: maybe @jmartinez-silabs could create a patch on the TE4 branch or asking some help from @vivien-apple / @bzbarsky-apple if they are more familiar with the regen (Thinking that this may be as easy as "cherrypick and run zap regen, however I am afraid that I cannot validate the result myself)

@franck-apple FIY

**Comment by andy31415** [comment]
We also have the option to pull #8065 in since that seems to have moved files around. That would probably resolve this merge conflict but unsure about any future issues.

**Comment by andy31415** [comment]
I cherrypicked and done the following to resolve conflicts:

```
cd third_party/zap/repo
npm install
cd -
./scripts/tools/zap_regen_all.py
```

and added/updated all files

**Comment by jmartinez-silabs** [comment]
@andy31415 Thank you Andrei

**Pull request merged**
